---
title: "BCL project report - ONT data"
output:
 html_document:
  highlight: tango
  theme: cosmo
  toc: no
  css: custom.css
params:
 path:
  label: "Path to ONT run folder (required, absolute path or relative to current folder)"
  value: "/path/to/folder"
  input: text
 sequencer:
  label: "Sequencing platform"
  choices: ["MinION", "GridION", "PromethION"]
  value: "PromethION"
  input: select
   
---

Report generated on `r Sys.time()`. All fastq data is calculated with the [faster](https://github.com/angelovangel/faster) and [fastkmers](https://github.com/angelovangel/fastkmers) programs. 


```{r setup, include=FALSE}
knitr::opts_chunk$set(include = FALSE, 
                      echo = FALSE, 
                      warning = FALSE, 
                      cache = FALSE)
#require(seqTools)
require(stringr)
require(knitr)
require(DT)
#require(kableExtra)
require(dplyr)
require(parallel)


fastqdir <- normalizePath(params$path)
fastqfiles <- list.files(fastqdir, pattern = "fast(q|q.gz)$", full.names = TRUE, recursive = params$recursive)
names(fastqfiles) <- basename(fastqfiles)
numcores <- ifelse(parallel::detectCores() > 2, parallel::detectCores(), 2)

```

```{r faster_table, message=FALSE}

stats_headers <- c("file", "num_reads", "bases", "n_bases", 
									 "min_len", "max_len", "avg_len", "Q1", "Q2", "Q3", 
									 "N50", "Q20_percent", "Q30_percent")

stats_command <- function(x) {
  system2(command = "faster", 
          args = c("-ts", x),
          stdout = TRUE)
  }

message(paste0('Running faster with ', numcores, ' cores...'))
stats <- parallel::mcmapply(stats_command, fastqfiles, mc.cores = numcores, mc.preschedule = T)
#stats <- parallel::mcmapply(stats_command, fastqfiles)

df <- stats %>%
	read.table(text = ., col.names = stats_headers) %>% 
	dplyr::mutate(file = basename(file)) %>% 
	dplyr::arrange(file) %>%
	dplyr::select(-c(8:10))



# 
write.csv(df, file = "stats.csv", row.names = FALSE)

# get some summary data
num_files <- length(fastqfiles)
num_reads <- df %>% dplyr::summarise(seqs = sum(num_reads)) %>% as.numeric()
num_reads_text <- system2("bin/siformat.sh", args = num_reads, stdout = T)
num_bases <- df %>% dplyr::summarise(bases = sum(bases)) %>% as.numeric()
num_bases_text <- system2("bin/siformat.sh", args = num_bases, stdout = T)
```

***

### Number of reads and read quality metrics
This is **`r params$sequencer`** sequencing data. The report contains 
<blockquote>  
**`r num_files`** fastq files  
**`r num_reads_text`** reads    
**`r num_bases_text`** bases  
</blockquote>
```{r table1, include=TRUE}


DT::datatable(df,
							escape = F,
							filter = 'top', 
							caption = paste0(""),
					extensions = 'Buttons', 
					options = list(dom = 'Btp', paging = FALSE,
												 buttons = c('copy', 'csv', 'excel')
												 ), 
					rownames = FALSE, 
					class = 'hover row-border') %>%
	DT::formatRound(2:7, 0) %>%
	DT::formatRound(8:9, 2)

```

***
